[Connector] Adding internal route for requesting ad-hoc ServiceNow access token

ymao1 opened the pull request:
Towards https://github.com/elastic/kibana/issues/130306

## Summary

Addresses item 6 from https://github.com/elastic/kibana/issues/130306: `Expose new actions API HTTP route, which allow to fetch an access token for not saved connectors.`

This PR adds an internal route to request an ad-hoc access token from a ServiceNow instance without requiring a saved connector. This allows the Create ServiceNow connector UI to request information from the `api/x_elas2_inc_int/elastic_api/health` endpoint during the ServiceNow connector creation process.

## To Verify
1. Run this PR and submit the following POST request:

```
POST https://localhost:5601/internal/actions/connector/_servicenow_access_token
{
  "apiUrl": "https://ven04334.service-now.com/",
  "config": {
    "clientId": <clientId>,
    "userIdentifierValue": "elastic_integration",
    "jwtKeyId": <jwtKeyId>
  },
  "secrets": {
    "clientSecret": <clientSecret>,
    "privateKey": <privateKey>
  }
}
```

2. Verify you get an access token in response in the form: `Bearer <accessToken>`


### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

ersin-erdal reviewed (approved):
LGTM

ymao1 commented:
@elasticmachine merge upstream

ymao1 commented:
@jportner We are adding this internal endpoint for generating ad-hoc access tokens for querying the health API for ServiceNow instances. Wanted to make sure @elastic/kibana-security was aware and wondering if there are any concerns y'all might have.

cnasikas commented on x-pack/plugins/actions/server/routes/get_service_now_access_token.ts:
```diff
@@ -0,0 +1,66 @@
+) => {
+  router.post(
+    {
+      path: `${INTERNAL_BASE_ACTION_API_PATH}/connector/_servicenow_access_token`,
```
Is it possible to generalize the endpoint so it can be used by all connectors for future usage?

ymao1 replied:
I pushed a commit to refactor some of the utility functions to be more generic.

cnasikas replied:
Thanks for the changes! Sorry for the extra work. I thought it was simpler.

cnasikas commented on x-pack/test/alerting_api_integration/security_and_spaces/tests/actions/builtin_action_types/servicenow_access_token.ts:
```diff
@@ -0,0 +1,73 @@
+    });
+
+    it('should return 200 when requesting an access token with OAuth credentials', async () => {
+      const { privateKey } = generateKeyPairSync('rsa', {
```
Given that this is a somehow expensive task to have on the CI, can we use a fixed one? Is the key from here `packages/kbn-dev-utils/src/certs.ts` suitable for our case? You can import it as `import { KBN_KEY_PATH } from '@kbn/dev-utils'`

ymao1 replied:
Great idea! I didn't know about that so thank you for pointing it out. Updated in this commit: https://github.com/elastic/kibana/pull/131171/commits/d3422a09f2f0429614ce83af33bf1da0ad21e085

ymao1 commented:
@cnasikas @ersin-erdal I pushed a refactoring to move the functions for getting OAuth access tokens into shared library functions and made it more generic. Now we have functions to get a JWT OAuth access token and a Client Credentials OAuth access token and the route allows you to specify which type you want.

ymao1 commented:
@elasticmachine merge upstream

mikecote commented on x-pack/plugins/actions/server/routes/get_oauth_access_token.ts:
```diff
@@ -0,0 +1,115 @@
+    router.handleLegacyErrors(
+      verifyAccessAndContext(licenseState, async function (context, req, res) {
+        const { type, options } = req.body;
+
```
Should we limit access to this API to users who can edit connectors? I realize in the current state, non-Kibana users can access this endpoint and create tokens.

From the pattern in actions client (`await this.authorization.ensureAuthorized('update');`) should we also audit log this? cc @kobelb the API creates tokens detached from a connector, should we track an audit of these API calls somehow?

jportner replied:
> Should we limit access to this API to users who can edit connectors? I realize in the current state, non-Kibana users can access this endpoint and create tokens.

Good idea, I agree we should limit access in this way :+1:

> From the pattern in actions client (`await this.authorization.ensureAuthorized('update');`) should we also audit log this? cc @kobelb the API creates tokens detached from a connector, should we track an audit of these API calls somehow?

I'm a bit on the fence on this --
On one hand, it would be great to audit anything and everything that happens in Kibana 🎉 

On the other hand:
1. I question how useful a specific audit record for this would be. We already have audit records for HTTP requests, so we know user X made a request to this API endpoint. The only additional thing a new specific audit event would give an administrator is the ability to determine what was in the payload (`tokenUrl`, `scope`, etc.)
2. I'm not sure what "Category" such a new audit event would fit into. Our current audit event categories are "authentication", "database", and "web".

I propose that we rely on the existing HTTP request auditing for this event, and we just add a regular log message with a "Debug" log level that contains the non-secret config values in the request payload (`tokenUrl`, `scope`, etc) and the outcome (success or fail).

ymao1 replied:
@mikecote @jportner Restricted access to users with ability to edit connectors in this commit: https://github.com/elastic/kibana/pull/131171/commits/79962f5e4bbc1564ffb9e1fd40b2d94c42a7b46b

Also debug logging the success/failure of token generation.

mikecote replied:
Thank you! Commit LGTM!

mikecote reviewed (commented):
I finished my review, the overall structure looks great but I have two questions before giving it a 👍

jportner commented:
> @jportner We are adding this internal endpoint for generating ad-hoc access tokens for querying the health API for ServiceNow instances. Wanted to make sure @elastic/kibana-security was aware and wondering if there are any concerns y'all might have.

Thanks for the ping, I'll review today

jportner commented on x-pack/plugins/actions/server/routes/get_oauth_access_token.ts:
```diff
@@ -0,0 +1,121 @@
+      verifyAccessAndContext(licenseState, async function (context, req, res) {
+        const { type, options } = req.body;
+
+        try {
+          configurationUtilities.ensureUriAllowed(options.tokenUrl);
+        } catch (err) {
+          return res.badRequest({ body: err.message });
+        }
```
Above and beyond this hostname validation, can we check to ensure that the URL contains _a hostname_ (it should never be a local path), and can we validate the URL scheme to the minimum that's needed (we should only need `https` right?)

See [this awesome blog post](https://www.acunetix.com/blog/articles/server-side-request-forgery-vulnerability/) for more information about mitigating SSRF.

ymao1 replied:
Added the recommended additional validation in this commit: https://github.com/elastic/kibana/pull/131171/commits/52b94de3fa3474a5a4c21b604335195007c66677

jportner reviewed (commented):
I responded to one of Mike's comments, other than that I don't have any concerns about this PR.

I think the overall approach makes sense :+1:

There's an SSRF risk here, but I think it's pretty minimal, with the combination of:
* Authorization checks (as suggested in one of the comments above)
* Additional URL validation (as suggested in my comment below)
* The optional hostname allow-list
* The inability for users to specify arbitrary parameters with the token request
* The logic which does not return the response body to users, it parses the response from the remote server and only returns the access token to users

In reviewing this, I noticed a couple small problems outside of the scope of this PR:

### 1. URL validation

Here's the implementation for the `tokenUrl` validation:

https://github.com/elastic/kibana/blob/0998b67ded645b1e12ee7e9c6181bb54700e5425/x-pack/plugins/actions/server/actions_config.ts#L70-L84

This code ignores _network-path references_ (see [URL Confusion Vulnerabilities](https://snyk.io/blog/url-confusion-vulnerabilities/)). In other words, a URL of `//my-domain.com/foo` would be parsed and the validation function would treat as if it has no hostname, but it _actually_ has a hostname of `my-domain.com` (in terms of how HTTP clients typically behave).

Fortunately the validation function fails securely; if an allow-list is configured and a URL doesn't contain a hostname, it fails to validate. However, I think you should change the usage of [`url.parse` to include the `slashesDenoteHost` parameter](https://nodejs.org/api/url.html#urlparseurlstring-parsequerystring-slashesdenotehost):

```diff
diff --git a/x-pack/plugins/actions/server/actions_config.ts b/x-pack/plugins/actions/server/actions_config.ts
index 35e08bb5cfe..49f1d1fd544 100644
--- a/x-pack/plugins/actions/server/actions_config.ts
+++ b/x-pack/plugins/actions/server/actions_config.ts
@@ -76,7 +76,7 @@ function isAllowed({ allowedHosts }: ActionsConfig, hostname: string | null): bo
 
 function isHostnameAllowedInUri(config: ActionsConfig, uri: string): boolean {
   return pipe(
-    tryCatch(() => url.parse(uri)),
+    tryCatch(() => url.parse(uri, false /* parseQueryString */, true /* slashesDenoteHost */)),
     map((parsedUrl) => parsedUrl.hostname),
     mapNullable((hostname) => isAllowed(config, hostname)),
     getOrElse<boolean>(() => false)
```

### 2. JWT assertions

The code to create JWT assertions allows for consumers to specify custom claims that would override the existing claims (subject, audience, issuer, etc.):

https://github.com/elastic/kibana/blob/53272c6c0e650a5aba1fe49b4c3b67c20a575be6/x-pack/plugins/actions/server/builtin_action_types/lib/create_jwt_assertion.ts#L19-L38

It doesn't appear that this `customClaims` option is being used anywhere. I'm not sure why it was added, but IMO if you want to keep it you should ensure that it can't be used as a vector to overwrite the other claims:

```diff
diff --git a/x-pack/plugins/actions/server/builtin_action_types/lib/create_jwt_assertion.ts b/x-pack/plugins/actions/server/builtin_action_types/lib/create_jwt_assertion.ts
index b33a2d17ed9..2fcb17c1e18 100644
--- a/x-pack/plugins/actions/server/builtin_action_types/lib/create_jwt_assertion.ts
+++ b/x-pack/plugins/actions/server/builtin_action_types/lib/create_jwt_assertion.ts
@@ -29,12 +29,12 @@ export function createJWTAssertion(
   const headerObj = { algorithm: 'RS256' as Algorithm, ...(keyId ? { keyid: keyId } : {}) };
 
   const payloadObj = {
+    ...(customClaims ?? {}),
     sub: subject, // subject claim identifies the principal that is the subject of the JWT
     aud: audience, // audience claim identifies the recipients that the JWT is intended for
     iss: issuer, // issuer claim identifies the principal that issued the JWT
     iat, // issued at claim identifies the time at which the JWT was issued
     exp: iat + (expireInMilliseconds ?? 3600), // expiration time claim identifies the expiration time on or after which the JWT MUST NOT be accepted for processing
-    ...(customClaims ?? {}),
   };
 
   try {
```

This will prevent us from accidentally introducing a vulnerability in the future 😄

ymao1 commented:
@jportner Thanks for reviewing! I've pushed changes based on your feedback:

> ### 1. URL validation

Added `slashesDenoteHost` in this commit: https://github.com/elastic/kibana/pull/131171/commits/769176f3824477365a67e9bee8bf69d8e069b90b

> ### 2. JWT assertions

Removed ability to specify `customClaims` in this commit: https://github.com/elastic/kibana/pull/131171/commits/b02b197360bcf40791cae4b001069a39f6439205. As you said, we are not currently using this so it makes sense to remove and re-add if the need ever comes up in the future.

mikecote reviewed (approved):
Changes LGTM!

jportner reviewed (approved):
LGTM, thanks for the quick turnaround!
